File: client/javascript/src/mpc/eden.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {Inspector, spi, Type} from "@/macro";
import {Context} from "@/prsim";
import {Reference, Service} from "@/kinds";
import {Execution} from "@/mpc/execution";

@spi("mesh")
export abstract class Eden {

    /**
     * Define the reference object.
     * @param ctx
     * @param metadata
     * @param reference
     */
    abstract define(ctx: Context, metadata: any, reference: Type<any>): any;

    /**
     * Refer the service reference by method.
     * @param ctx
     * @param metadata
     * @param reference
     * @param method
     */
    abstract refer(ctx: Context, metadata: any, reference: Type<any>, method: Inspector): Execution<Reference>;

    /**
     * Store the service object.
     * @param ctx
     * @param kind
     * @param service
     */
    abstract store(ctx: Context, kind: Type<any>, service: any): void;

    /**
     * Infer the reference service by domain.
     * @param ctx
     * @param urn
     */
    abstract infer(ctx: Context, urn: string): Execution<Service>;

    /**
     * Get all reference types.
     * @param ctx
     */
    abstract referTypes(ctx: Context): Type<any>[];

    /**
     * Get all service types.
     * @param ctx
     */
    abstract inferTypes(ctx: Context): Type<any>[];
}

File: client/golang/mpc/eden.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package mpc

import (
	"context"
	"github.com/opendatav/mesh/client/golang/macro"
)

var IEden = (*Eden)(nil)

// Eden
// @SPI("mesh")
type Eden interface {

	// Define the reference object.
	Define(ctx context.Context, mpi macro.MPI, reference interface{}) error

	// Refer the service reference by method.
	Refer(ctx context.Context, mpi macro.MPI, reference interface{}, method macro.Inspector) (Execution, error)

	// Store the service object.
	Store(ctx context.Context, kind interface{}, service interface{}) error

	// Infer the reference service by Domain.
	Infer(ctx context.Context, urn string) (Execution, error)

	// ReferTypes Get all reference types.
	ReferTypes(ctx context.Context) ([]macro.MPI, error)

	// InferTypes Get all service types.
	InferTypes(ctx context.Context) ([]macro.MPS, error)
}
